### PR TITLE
879 NDM proxy

### PR DESCRIPTION
**Pull request by spetz**

### CODE REVIEW

File: src/Nethermind/Nethermind.Config/ConfigSourceHelper.cs [hunk 1]
```diff
@@ -101,7 +101,13 @@ private static object GetValue(Type valueType, string itemValue)
 
                 throw new IOException($"Cannot parse enum value: {itemValue}, type: {valueType.Name}");
             }
-            
+
+            if (valueType == typeof(string) && itemValue.StartsWith("[") && itemValue.EndsWith("]") &&
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Config/ConfigSourceHelper.cs]: here we just follow the bad practices

File: src/Nethermind/Nethermind.DataMarketplace.Consumers/Deposits/Domain/DepositDetails.cs [hunk 2]
```diff
@@ -72,6 +72,11 @@ public void SetConfirmationTimestamp(uint timestamp)
             ConfirmationTimestamp = timestamp;
         }
 
+        public void SetTransactionHash(Keccak transactionHash)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.DataMarketplace.Consumers/Deposits/Domain/DepositDetails.cs]: why not setter here? looks strange, like it was something different and then back simple again

File: src/Nethermind/Nethermind.DataMarketplace.Consumers/Deposits/Services/DepositConfirmationService.cs [hunk 3]
```diff
@@ -126,15 +124,16 @@ public async Task TryConfirmAsync(DepositDetails deposit)
                     break;
                 }
 
-                if (receipt.BlockHash == block.Hash || block.Number <= receipt.BlockNumber)
+                if (transaction.BlockHash == block.Hash || block.Number <= transaction.BlockNumber)
                 {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.DataMarketplace.Consumers/Deposits/Services/DepositConfirmationService.cs]: better to make this check before calling blockchain bridge

File: src/Nethermind/Nethermind.DataMarketplace.Consumers/Receipts/Services/ReceiptService.cs [hunk 4]
```diff
@@ -74,34 +73,13 @@ public class ReceiptService : IReceiptService
             _logger = logManager.GetClassLogger();
         }
         
-        public async Task SendAsync(DataDeliveryReceiptRequest request)
+        public async Task SendAsync(DataDeliveryReceiptRequest request, int fetchSessionRetries = 3,
+            int fetchSessionRetryDelayMilliseconds = 3000)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.DataMarketplace.Consumers/Receipts/Services/ReceiptService.cs]: are the delays and retries tested?

File: src/Nethermind/Nethermind.DataMarketplace.Core/Services/TransactionVerifier.cs [hunk 5]
```diff
@@ -16,42 +16,46 @@
...
 
-        public TransactionVerifier(IBlockchainBridge blockchainBridge, uint requiredBlockConfirmations)
+        public TransactionVerifier(INdmBlockchainBridge blockchainBridge, uint requiredBlockConfirmations)
         {
             _blockchainBridge = blockchainBridge;
             _requiredBlockConfirmations = requiredBlockConfirmations;
         }
         
-        public TransactionVerifierResult Verify(TxReceipt receipt)
+        public async Task<TransactionVerifierResult> VerifyAsync(NdmTransaction transaction)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.DataMarketplace.Core/Services/TransactionVerifier.cs]: I wonder if you take into account that transaction.BlockNumber may change

File: src/Nethermind/Nethermind.Facade/Proxy/DefaultHttpClient.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+using System.Diagnostics;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Nethermind.Core;
+using Nethermind.Logging;
+
+namespace Nethermind.Facade.Proxy
+{
+    public class DefaultHttpClient : IHttpClient
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/DefaultHttpClient.cs]: have you reviewed if can remove any existing implementations of IJsonRpcClient?

File: src/Nethermind/Nethermind.Facade/Proxy/Models/BlockModel.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Dirichlet.Numerics;
+
+namespace Nethermind.Facade.Proxy.Models
+{
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/Models/BlockModel.cs]: why not using BlockForRpc here?

File: src/Nethermind/Nethermind.Facade/Proxy/Models/BlockParameterModel.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using Nethermind.Dirichlet.Numerics;
+
+namespace Nethermind.Facade.Proxy.Models
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/Models/BlockParameterModel.cs]: why not using existing types from RPC?

File: src/Nethermind/Nethermind.Facade/Proxy/Models/CallTransactionModel.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using Nethermind.Core;
+using Nethermind.Dirichlet.Numerics;
+
+namespace Nethermind.Facade.Proxy.Models
+{
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/Models/CallTransactionModel.cs]: why not the types from JSON RPC?

File: src/Nethermind/Nethermind.Facade/Proxy/Models/LogModel.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Dirichlet.Numerics;
+
+namespace Nethermind.Facade.Proxy.Models
+{
+    public class LogModel
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/Models/LogModel.cs]: WHy not LogForRpc?

File: src/Nethermind/Nethermind.Facade/Proxy/Models/ReceiptModel.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Dirichlet.Numerics;
+
+namespace Nethermind.Facade.Proxy.Models
+{
+    public class ReceiptModel
+    {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/Models/ReceiptModel.cs]: WHy not ReceiptForRpc?

File: src/Nethermind/Nethermind.Facade/Proxy/Models/TransactionModel.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        public UInt256 BlockNumber { get; set; }
+        public Address From { get; set; }
+        public Address To { get; set; }
+        public UInt256 Gas { get; set; }
+        public UInt256 GasPrice { get; set; }
+        public byte[] Input { get; set; }
+        public UInt256 Value { get; set; }
+
+        public Transaction ToTransaction()
+            => new Transaction
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/Models/TransactionModel.cs]: WHy not TransactionForRpc?

File: src/Nethermind/Nethermind.Facade/Proxy/RpcResult.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+namespace Nethermind.Facade.Proxy
+{
+    public class RpcResult<T>
+    {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Facade/Proxy/RpcResult.cs]: why not using the existing solution?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthModuleProxy.cs [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        private readonly IEthJsonRpcClientProxy _proxy;
+
+        public EthModuleProxy(IEthJsonRpcClientProxy proxy)
+        {
+            _proxy = proxy;
+        }
+
+        public ResultWrapper<string> eth_protocolVersion()
+        {
+            throw new System.NotImplementedException();
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthModuleProxy.cs]: in such cases better use NotSupportedException maybe?

File: src/Nethermind/Nethermind.Runner/configs/ndm_consumer.cfg [hunk 15]
```diff
@@ -1,36 +1,24 @@
 {
   "Init": {
     "EnableUnsecuredDevWallet": true,
-    "KeepDevWalletInMemory": true,
+    "KeepDevWalletInMemory": false,
     "WebSocketsEnabled": true,
-    "NetworkEnabled": true,
-    "DiscoveryEnabled": false,
+    "StoreReceipts": true,
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Runner/configs/ndm_consumer.cfg]: Did you add tests for it?

File: src/Nethermind/Nethermind.Runner/configs/ndm_consumer_local.cfg [hunk 16]
```diff
@@ -3,31 +3,26 @@
     "EnableUnsecuredDevWallet": true,
     "KeepDevWalletInMemory": false,
     "WebSocketsEnabled": true,
-    "DiscoveryEnabled": false,
+    "StoreReceipts": true,
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Runner/configs/ndm_consumer_local.cfg]: why enabling discovery on local setups?

Comment (spetz) [reply src/Nethermind/Nethermind.Runner/configs/ndm_consumer_local.cfg]: These are removed lines.

File: src/Nethermind/Nethermind.Facade/Proxy/RpcResult.cs (hunk 13 above)
Comment (spetz) [reply src/Nethermind/Nethermind.Facade/Proxy/RpcResult.cs]: `Nethermind.JsonRpc` already contains a reference to `Nethermind.Facade` and the Proxy models are just plain DTOs without any additional processing logic whatsoever.

File: src/Nethermind/Nethermind.Runner/configs/ndm_consumer.cfg (hunk 15 above)
Comment (spetz) [reply src/Nethermind/Nethermind.Runner/configs/ndm_consumer.cfg]: Tests for? It was just an updated config, similarly to `ndm_consumer_goerli.cfg` etc.

**Pull request merged**